### PR TITLE
Refactor GameRunner

### PR DESCRIPTION
**Pull request by DanVanAtta**
- Migrate code from HeadlessGameServer (bot) to GameRunner. Do so by moving code from one to the other, and forking the logic depending upon the value of an enum. After doing this, work to consolidate the two code paths.

- Thorough cleanup of GameRunner: Some more important changes is the carving off of classes to handle some of the static functionality in GameRunner. Perhaps about 70% of that effort is complete, there is more decomposition to submodules to be done in GameRunner.

- I got pretty deep in the background task runner:
 - The latch stuff made it blocking, that was removed, then things were fixed to handle the fact it was no longer blocking
 - The usage of background task runner was reduced. Before we had two places, when checking for out dated maps, and second when downloading the download file list. The first is done in a background thread anyways, so we do not need to show the foreground spinner, simply show the 'do you want to update maps?' dialog when needed. Now that it is less annoying, we can consider increasing the frequency of the check.

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/1046)

### CODE REVIEW

File: src/games/strategy/debug/DebugUtils.java [hunk 1]
```diff
@@ -70,7 +71,7 @@ public static String getMemory() {
     buf.append("Free memory: ").append(runtime.freeMemory() / mb).append("\r\n");
     buf.append("Total memory: ").append(runtime.totalMemory() / mb).append("\r\n");
     buf.append("Max memory: ").append(runtime.maxMemory() / mb).append("\r\n");
-    final int currentMaxSetting = GameRunner.getMaxMemoryFromSystemIniFileInMB(GameRunner.getSystemIni());
+    final int currentMaxSetting = Memory.getMaxMemoryFromSystemIniFileInMB(GameRunner.getSystemIni());
```
Comment (DanVanAtta) [inline src/games/strategy/debug/DebugUtils.java]: GameRunner has a lot of static 'stuff', you'll see a pattern of this moving out of GameRunner.

File: src/games/strategy/engine/framework/GameDataManager.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public GameData loadGame(final ObjectInputStream input, final String savegamePat
         // same engine as was
         // used for this savegame, and if so try to run it
         try {
-          final String newClassPath = TripleAProcessRunner.findOldJar(readVersion, true);
+          final String newClassPath = GameRunner.findOldJar(readVersion, true);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameDataManager.java]: This on the other hand is the oppositie, TripleAProcessor has a lot of static functions and is closely related to the internals of GameRunner. Consolidating the two allowed more variables to be private between the two classes.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 3]
```diff
@@ -50,14 +52,18 @@
  */
 public class GameRunner {
 
+  public enum GameMode { SWING_CLIENT, HEADLESS_BOT }
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Important enum, the headless server code was copy pasted here and then put behind this enum. After that we simplify.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 4]
```diff
@@ -50,14 +52,18 @@
...
+
+  public static final String TRIPLEA_HEADLESS = "triplea.headless";
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: All props from the headless server were moved to GameRunner, consolidating to GameRunner so it has teh responsibility of launching games.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 5]
```diff
@@ -76,96 +82,103 @@
...
   public static final int MINIMUM_SERVER_START_GAME_SYNC_WAIT_TIME =
       MINIMUM_SERVER_OBSERVER_JOIN_WAIT_TIME + ADDITIONAL_SERVER_ERROR_DISCONNECTION_WAIT_TIME + 110;
-  public static final int DEFAULT_SERVER_START_GAME_SYNC_WAIT_TIME =
+  private static final int DEFAULT_SERVER_START_GAME_SYNC_WAIT_TIME =
       Math.max(Math.max(MINIMUM_SERVER_START_GAME_SYNC_WAIT_TIME, 900),
           DEFAULT_SERVER_OBSERVER_JOIN_WAIT_TIME + ADDITIONAL_SERVER_ERROR_DISCONNECTION_WAIT_TIME + 110);
 
   private static final String MAP_FOLDER = "mapFolder";
 
+  private static String[] COMMAND_LINE_ARGS =
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Before this array was returned by a method, the removal is later on, and we have the consolidated args from the bot version.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 6]
```diff
@@ -76,96 +82,103 @@
...
+      {TRIPLEA_GAME_PROPERTY, TRIPLEA_SERVER_PROPERTY, TRIPLEA_CLIENT_PROPERTY, TRIPLEA_HOST_PROPERTY,
+          TRIPLEA_PORT_PROPERTY, TRIPLEA_NAME_PROPERTY, TRIPLEA_SERVER_PASSWORD_PROPERTY, TRIPLEA_STARTED,
+          LobbyServer.TRIPLEA_LOBBY_PORT_PROPERTY,
+          LOBBY_HOST, LOBBY_GAME_COMMENTS, LOBBY_GAME_HOSTED_BY, TRIPLEA_ENGINE_VERSION_BIN, HttpProxy.PROXY_HOST,
+          HttpProxy.PROXY_PORT, TRIPLEA_DO_NOT_CHECK_FOR_UPDATES, Memory.TRIPLEA_MEMORY_SET, MAP_FOLDER};
+
+
+  private static void usage(GameMode gameMode) {
+    if(gameMode == GameMode.HEADLESS_BOT) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: The bot or swing client fork.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 7]
```diff
@@ -76,96 +82,103 @@
...
+      System.out.println("\nUsage and Valid Arguments:\n"
+          + "   " + TRIPLEA_GAME_PROPERTY + "=<FILE_NAME>\n"
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: usage is now formatted : )
 not consolidated, but it is easier to manage them both when they are in one spot.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 8]
```diff
@@ -76,96 +82,103 @@
...
+          + "   " + LOBBY_HOST + "=<LOBBY_HOST>\n"
+          + "   " + LobbyServer.TRIPLEA_LOBBY_PORT_PROPERTY + "=<LOBBY_PORT>\n"
+          + "   " + LOBBY_GAME_COMMENTS + "=<LOBBY_GAME_COMMENTS>\n"
+          + "   " + LOBBY_GAME_HOSTED_BY + "=<LOBBY_GAME_HOSTED_BY>\n"
+          + "   " + LOBBY_GAME_SUPPORT_EMAIL + "=<youremail@emailprovider.com>\n"
+          + "   " + LOBBY_GAME_SUPPORT_PASSWORD + "=<password for remote actions, such as remote stop game>\n"
+          + "   " + LOBBY_GAME_RECONNECTION + "=<seconds between refreshing lobby connection [min " + LOBBY_RECONNECTION_REFRESH_SECONDS_MINIMUM + "]>\n"
+          + "   " + TRIPLEA_SERVER_START_GAME_SYNC_WAIT_TIME + "=<seconds to wait for all clients to start the game>\n"
+          + "   " + TRIPLEA_SERVER_OBSERVER_JOIN_WAIT_TIME + "=<seconds to wait for an observer joining the game>\n"
+          + "   " + MAP_FOLDER + "=mapFolder"
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: This arg quietly was moved to the bot part of this fork.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 9]
```diff
@@ -177,37 +190,28 @@ private static void showMainFrame() {
...
-          startsWithPropertyKey = true;
-          break;
-        }
-      }
-      if (!startsWithPropertyKey) {
-        // change it to start with the key
-        args[0] = TRIPLEA_GAME_PROPERTY + "=" + args[0];
-      }
+  public static void handleCommandLineArgs(final String[] args, final String[] availableProperties, GameMode gameMode) {
+    if (args.length == 1 && !argsStartsWithPropertykey(args[0])) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: The value of `argsStartsWithPropertykey` was extracted to a  method, simplified the block compared to to the roughly 7 lines it took before.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 10]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
-        // if successful we don't do anything
-        System.out.println(TRIPLEA_ENGINE_VERSION_BIN + ":" + version);
-        if (!ClientContext.engineVersion().getVersion().equals(testVersion, false)) {
-          System.out.println("Current Engine version in use: " + ClientContext.engineVersion());
-        }
-      } catch (final Exception e) {
-        System.getProperties().setProperty(TRIPLEA_ENGINE_VERSION_BIN, ClientContext.engineVersion().toString());
-        System.out.println(TRIPLEA_ENGINE_VERSION_BIN + ":" + ClientContext.engineVersion());
+
+    if (gameMode == GameMode.HEADLESS_BOT) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: The diff gets pretty hard to follow here. The changes in this section are more easily reviewed commit-by-commit.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 11]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
-  public static long getMaxMemoryInBytes() {
-    final Properties systemIni = getSystemIni();
-    final String useDefaultMaxMemoryString = systemIni.getProperty(TRIPLEA_MEMORY_USE_DEFAULT, "true");
-    final boolean useDefaultMaxMemory = Boolean.parseBoolean(useDefaultMaxMemoryString);
-    final String maxMemoryString = systemIni.getProperty(TRIPLEA_MEMORY_XMX, "").trim();
-    // for whatever reason, .maxMemory() returns a value about 12% smaller than the real Xmx value.
-    // Just something to be aware of.
-    long max = Runtime.getRuntime().maxMemory();
-    if (!useDefaultMaxMemory && maxMemoryString.length() > 0) {
+  public static void setupLogging(GameMode gameMode) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: setupLogging was also in the headless version of GameRunner. Same fork strategy was applied here. Diff is still a bit difficult to decipher..

File: src/games/strategy/engine/framework/GameRunner.java [hunk 12]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
-      } catch (final NumberFormatException e) {
+        LogManager.getLogManager().readConfiguration(ClassLoader.getSystemResourceAsStream("logging.properties"));
+      } catch (final Exception e) {
         ClientLogger.logQuietly(e);
       }
-    }
-    return max;
-  }
-
-  public static int getMaxMemoryFromSystemIniFileInMB(final Properties systemIni) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: memory stuff was moved to Memory.java

File: src/games/strategy/engine/framework/GameRunner.java [hunk 13]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
-    final String forOnlineOnlyString = systemIni.getProperty(TRIPLEA_MEMORY_ONLINE_ONLY, "true");
-    final boolean forOnlineOnly = Boolean.parseBoolean(forOnlineOnlyString);
-    return forOnlineOnly;
-  }
 
   public static Properties getSystemIni() {
     final Properties rVal = new Properties();
     final File systemIni = new File(ClientFileSystemHelper.getRootFolder(), SYSTEM_INI);
-    if (systemIni != null && systemIni.exists()) {
+    if (systemIni.exists()) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: systemIni will never be null

File: src/games/strategy/engine/framework/GameRunner.java [hunk 14]
```diff
@@ -437,15 +397,12 @@ public static Properties getSystemIni() {
     return rVal;
   }
 
-  public static void writeSystemIni(final Properties properties, final boolean clearOldAndOverwrite) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: value of this arg was always false

File: src/games/strategy/engine/framework/GameRunner.java [hunk 15]
```diff
@@ -457,153 +414,6 @@ public static void writeSystemIni(final Properties properties, final boolean cle
     }
   }
 
-  private static void setupProxies() {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: moved to HttpProxy.java

File: src/games/strategy/engine/framework/GameRunner.java [hunk 16]
```diff
@@ -635,13 +445,7 @@ public static boolean getCasualtySelectionSlow() {
...
 
   public static void setCasualtySelectionSlow(final boolean casualtySelectionBeta) {
-    final Preferences pref = Preferences.userNodeForPackage(GameRunner.class);
-    pref.putBoolean(CASUALTY_SELECTION_SLOW, casualtySelectionBeta);
-    try {
-      pref.sync();
-    } catch (final BackingStoreException e) {
-      ClientLogger.logQuietly(e);
-    }
+    SystemPreferences.put(SystemPreferenceKey.CASUALTY_SELECTION_SLOW, casualtySelectionBeta);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: SystemPreferences replaces the code chunk above. We get the sync() and error handling for free, and we get type safety with an enum parameter instead of a string. The default value `casualtySelectionBeta` should stay the same, and fewer lines. There were a number of places where this lib was used.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 17]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
     }
     return img;
   }
+
+  public static void startNewTripleA(final Long maxMemory) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Begin the tacking on of stuff from TripleAProcessor. There were minimal modifications done after moving the code.

File: src/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 18]
```diff
@@ -61,32 +62,34 @@ public static Version getVersion(final File zipFile) {
...
   public static void showDownloadMapsWindow() {
-    showDownloadMapsWindow(null, Optional.empty());
+    showDownloadMapsWindow(Optional.empty());
   }
 
 
-  private static void showDownloadMapsWindow(final Component parent, Optional<String> mapName) {
-    final DownloadRunnable download = new DownloadRunnable(ClientContext.mapListingSource().getMapListDownloadSite());
+  private static void showDownloadMapsWindow(Optional<String> mapName) {
+    Runnable downloadAndShowWindow = () -> {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: This was an important change, to deal with BAckground task runner no longer blocking.
Basically we set the show download window action to happen at the end of the same thread that does the file list download. Contrast that to doing the file list download, blocking, then in the foreground thread we show the download window.

File: src/games/strategy/engine/framework/map/download/DownloadRunnable.java [hunk 19]
```diff
@@ -34,7 +36,11 @@ public DownloadRunnable(final String urlString) {
   }
 
   public List<DownloadFileDescription> getDownloads() {
-    return downloads;
+   if(downloads == null) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/DownloadRunnable.java]: bit of a paranoid NPE check, might not actually be necessary, but it enforces the assumptions that this will never return null.

File: src/games/strategy/engine/framework/map/download/MapDownloadController.java [hunk 20]
```diff
@@ -1,90 +1,58 @@
...
 public class MapDownloadController {
 
-  private static final String TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES = "triplea.lastCheckForMapUpdates";
   private final MapListingSource mapDownloadProperties;
 
   public MapDownloadController(final MapListingSource mapSource) {
     mapDownloadProperties = mapSource;
   }
 
-  public DownloadRunnable downloadForLatestMapsCheck() {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/MapDownloadController.java]: This method was inlined.

File: src/games/strategy/engine/framework/map/download/MapDownloadController.java [hunk 21]
```diff
@@ -1,90 +1,58 @@
...
-    final DownloadRunnable runnable = new DownloadRunnable(mapDownloadProperties.getMapListDownloadSite());
-    BackgroundTaskRunner.runInBackground(null, "Checking for out-of-date Maps.", runnable,
-        new CountDownLatchHandler(true));
-    return runnable;
-  }
-
-  public void downloadMap(String mapName) {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/MapDownloadController.java]: this method was unused, same with the one below

File: src/games/strategy/engine/framework/GameRunner.java (hunk 3 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: I would have used a boolean instead, but this works too I guess...

Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: That would have lead to boolean parameters, and then we have calls like this:
`setupLogging(false)`. Which is a good example of why boolean parameters are bad. There is really no good way to understand just reading that what it would mean. One tactic is to assign the `false` value to a local variable, but that is a more of a bandaid solution. Regardless, when you get in the staet that you have to jump through code to understand it, when you have to track down methods that are called to read them to understand it, it's not a good place. Imagine having to do that for 50 lines of code, tracing down a dozen methods, it would take forever to get through and to understand the code, and then in that state bugs become common since the code is often not well understood.

Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Totally agree - I figured this was for better understanding, just sayin...

**Comment by DanVanAtta** [comment]
This is needed for the 1.9 release - it allows the bot servers to be set up with a target maps folder.

File: src/games/strategy/debug/DebugUtils.java (hunk 1 above)
Comment (RoiEXLab) [inline src/games/strategy/debug/DebugUtils.java]: IMO the Method name is too long... It should be shortened to something like `getMaxMemoryFromFile` or even shorter... Further description should be moved to the javadoc

File: src/games/strategy/engine/framework/GameRunner.java (hunk 5 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Is this array used to determine if a given String is a valid argument?
If so, wouldn't it be easier to use a collection instead of an array? ('COMMAND_LINE_ARGS.contains("THIS STRING");')

File: src/games/strategy/engine/framework/GameRunner.java (hunk 9 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Is the parameter availableProperties really needed? The `COMMAND_LINE_ARGS` array is being passed in the main method

File: src/games/strategy/engine/framework/GameRunner.java [hunk 22]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
-    try {
-      pref.sync();
-    } catch (final BackingStoreException e) {
-      ClientLogger.logQuietly(e);
+  private static boolean argsStartsWithPropertykey(String firstArg) {
+    boolean startsWithPropertyKey = false;
+    for (final String arg : COMMAND_LINE_ARGS) {
+      if (firstArg.startsWith(arg)) {
+        startsWithPropertyKey = true;
+        break;
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Regarding the collection you could do a `forEach(e -> { if(firstArgs.startsWith(e)) return true;});` or something like that

File: src/games/strategy/engine/framework/GameRunner.java [hunk 23]
```diff
@@ -222,212 +226,168 @@ private static void handleCommandLineArgs(final String[] args) {
...
       }
-    }
-    return max;
-  }
-
-  public static int getMaxMemoryFromSystemIniFileInMB(final Properties systemIni) {
-    final String maxMemoryString = systemIni.getProperty(TRIPLEA_MEMORY_XMX, "").trim();
-    int maxMemorySet = -1;
-    if (maxMemoryString.length() > 0) {
+      Toolkit.getDefaultToolkit().getSystemEventQueue().push(new EventQueue() {
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Makes sense to move the UI related stuff in here

File: src/games/strategy/engine/framework/GameRunner.java (hunk 13 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: legit

File: src/games/strategy/engine/framework/GameRunner.java [hunk 24]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
...
+
+  public static void startGame(final String savegamePath, final String classpath, final Long maxMemory) {
+    final List<String> commands = new ArrayList<>();
+    if (maxMemory != null && maxMemory > (32 * 1024 * 1024)) {
+      ProcessRunnerUtil.populateBasicJavaArgs(commands, classpath, maxMemory);
+    } else {
+      ProcessRunnerUtil.populateBasicJavaArgs(commands, classpath);
+    }
+    if (savegamePath != null && savegamePath.length() > 0) {
+      commands.add("-D" + GameRunner.TRIPLEA_GAME_PROPERTY + "=" + savegamePath);
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: I can see lots of String being created like this... what about adding a static method like `generateArg(String arg, String value)`?

File: src/games/strategy/engine/framework/GameRunner.java [hunk 25]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
...
+    }
+    // add in any existing command line items
+    for (final String property : GameRunner.COMMAND_LINE_ARGS) {
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: That's what I mean... We are iterating through the Array instead of using a Collection

File: src/games/strategy/engine/framework/GameRunner.java [hunk 26]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
...
+    // classpath for main
+    commands.add(GameRunner.class.getName());
+    ProcessRunnerUtil.exec(commands);
+  }
+
+  public static void hostGame(final int port, final String playerName, final String comments, final String password,
+      final Messengers messengers) {
+    final List<String> commands = new ArrayList<>();
+    ProcessRunnerUtil.populateBasicJavaArgs(commands);
+    commands.add("-D" + TRIPLEA_SERVER_PROPERTY + "=true");
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Like I said: lots of `"-D" + key + "=" + value` Strings - easy to mistype

File: src/games/strategy/engine/framework/GameRunner.java [hunk 27]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
...
+    commands.add("-D" + LOBBY_GAME_COMMENTS + "=" + comments);
+    commands.add("-D" + LOBBY_GAME_HOSTED_BY + "=" + messengers.getMessenger().getLocalNode().getName());
+    if (password != null && password.length() > 0) {
+      commands.add("-D" + TRIPLEA_SERVER_PASSWORD_PROPERTY + "=" + password);
+    }
+    final String fileName = System.getProperty(TRIPLEA_GAME_PROPERTY, "");
+    if (fileName.length() > 0) {
+      commands.add("-D" + TRIPLEA_GAME_PROPERTY + "=" + fileName);
+    }
+    final String javaClass = "games.strategy.engine.framework.GameRunner";
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: This should be an unchecled String... this should be something like `GameRunner.class.getName() (not sure what the exact method is...)

File: src/games/strategy/engine/framework/GameRunner.java [hunk 28]
```diff
@@ -785,4 +574,217 @@ public static Image getGameIcon(final Window frame) {
...
+    if (ClientFileSystemHelper.areWeOldExtraJar()) {
+      final String version = System.getProperty(GameRunner.TRIPLEA_ENGINE_VERSION_BIN);
+      if (version != null && version.length() > 0) {
+        Version defaultVersion = null;
+        try {
+          defaultVersion = new Version(version);
+        } catch (final Exception e) {
+          // nothing, just continue
+        }
+        if (defaultVersion != null) {
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: if an exception happens the error should be printed... In this case if default version is null the code is not executed - it should be moved inside the try block

File: src/games/strategy/engine/framework/ProcessRunnerUtil.java [hunk 29]
```diff
@@ -35,7 +37,7 @@ public static void populateBasicJavaArgs(final List<String> commands, final Stri
     // add 64m to that to
     // compensate
     // final long maxMemory = ((long) (Runtime.getRuntime().maxMemory() * 1.15) + 67108864);
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/ProcessRunnerUtil.java]: Those comments should get removed.
I think the reason for th 12% drop is that the JVM requires some memory for itself...

File: src/games/strategy/engine/framework/map/download/DownloadRunnable.java (hunk 19 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/map/download/DownloadRunnable.java]: Why not initialize with `= new ArrayList<>();` ?

File: src/games/strategy/engine/framework/startup/ui/EnginePreferences.java [hunk 30]
```diff
@@ -221,18 +223,19 @@ private void setupListeners() {
           new CountDownLatchHandler(true));
     }));
     m_setupProxies.addActionListener(SwingAction.of("Setup Network and Proxy Settings", e -> {
+      // TODO: this action listener should probably come from the HttpProxy class
       final Preferences pref = Preferences.userNodeForPackage(GameRunner.class);
-      final ProxyChoice proxyChoice =
-          ProxyChoice.valueOf(pref.get(GameRunner.PROXY_CHOICE, ProxyChoice.NONE.toString()));
-      final String proxyHost = pref.get(GameRunner.PROXY_HOST, "");
+      final HttpProxy.ProxyChoice proxyChoice =
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/startup/ui/EnginePreferences.java]: I don't think the HttpProxy prefix is actually necessary

File: src/games/strategy/engine/framework/system/HttpProxy.java [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+      } catch (final BackingStoreException e) {
+        ClientLogger.logQuietly(e);
+      }
+    }
+  }
+
+  private static void setToUseSystemProxies() {
+    final String javaNetUseSystemProxies = "java.net.useSystemProxies";
+    System.setProperty(javaNetUseSystemProxies, "true");
+    List<java.net.Proxy> proxyList = null;
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/HttpProxy.java]: IMO `java.net.Proxy` should be imported

File: src/games/strategy/engine/framework/system/HttpProxy.java [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+    try {
+      final ProxySelector def = ProxySelector.getDefault();
+      if (def != null) {
+        proxyList = def.select(new URI("http://sourceforge.net/"));
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/HttpProxy.java]: Is there a sourceforge alternative?

File: src/games/strategy/engine/framework/system/HttpProxy.java [hunk 33]
```diff
@@ -0,0 +1,176 @@
...
+        ProxySelector.setDefault(null);
+        if (proxyList != null && !proxyList.isEmpty()) {
+          final java.net.Proxy proxy = proxyList.get(0);
+          final InetSocketAddress address = (InetSocketAddress) proxy.address();
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/HttpProxy.java]: I think using The abstract class SocketAdress is better than casting

File: src/games/strategy/engine/framework/system/Memory.java [hunk 34]
```diff
@@ -0,0 +1,119 @@
...
+    // if we have already set the memory, then return.
+    // (example: we used process runner to create a new triplea with a specific memory)
+    if (memSet) {
+      return;
+    }
+    final Properties systemIni = GameRunner.getSystemIni();
+    if (useDefaultMaxMemory(systemIni)) {
+      return;
+    }
+    if (getUseMaxMemorySettingOnlyForOnlineJoinOrHost(systemIni)) {
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/Memory.java]: Another one of those long method names... They are descriptive, but very hard to read

File: src/games/strategy/engine/framework/system/Memory.java [hunk 35]
```diff
@@ -0,0 +1,119 @@
...
+    final Properties systemIni = GameRunner.getSystemIni();
+    final String useDefaultMaxMemoryString = systemIni.getProperty(TRIPLEA_MEMORY_USE_DEFAULT, "true");
+    final boolean useDefaultMaxMemory = Boolean.parseBoolean(useDefaultMaxMemoryString);
+    final String maxMemoryString = systemIni.getProperty(TRIPLEA_MEMORY_XMX, "").trim();
+    // for whatever reason, .maxMemory() returns a value about 12% smaller than the real Xmx value.
+    // Just something to be aware of.
+    long max = Runtime.getRuntime().maxMemory();
+    if (!useDefaultMaxMemory && maxMemoryString.length() > 0) {
+      try {
+        final int maxMemorySet = Integer.parseInt(maxMemoryString);
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/Memory.java]: If maxMemorySet would be of type long, casting below wouldn't be necessary

File: src/games/strategy/engine/framework/system/SystemProperties.java [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * Wrapper class around System.getProperties(), use this class to set/get System properties.
+ * Prefer to use system props only for command line usage. TripleA code base has made pretty extensive use
+ * of System props to pass values, which is not a best practice. Converting those usages to this wrapper interface
+ * will make different parts of the code that use systemn properties much easier to manage.
+ */
+// final + private constructor to disallow inheritance, all access to the system is static
+public final class SystemProperties {
+
+  private enum SystemPropertyKey {
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/system/SystemProperties.java]: I can't see a use for a private enumType without values

File: src/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java [hunk 37]
```diff
@@ -1,75 +1,34 @@
...
-            }
+          // clean up the window
+          SwingUtilities.invokeLater(() -> {
+            doneWait.set(true);
+            window.setVisible(false);
+            window.dispose();
           });
         }
+      });
+      t.start();
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: the thread doesn't need to be saved as a local variable, it can be started directly: `new Thread().start();`

**Comment by RoiEXLab** [comment]
Looks good in general, after some things change it'll be fine

File: src/games/strategy/engine/framework/map/download/DownloadMapsWindow.java [hunk 38]
```diff
@@ -61,32 +62,34 @@ public static Version getVersion(final File zipFile) {
...
+      final DownloadRunnable download = new DownloadRunnable(ClientContext.mapListingSource().getMapListDownloadSite());
+      download.run();
+      final List<DownloadFileDescription> games = download.getDownloads();
+      checkNotNull(games);
```
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/map/download/DownloadMapsWindow.java]: @DanVanAtta This line is causing the merge conflict due to #1053

File: src/games/strategy/debug/DebugUtils.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/debug/DebugUtils.java]: I don't necessarily disagree. Not sure of a better name at the moment. Regardless, the method was moved out of this class to a new one. To some extent, the name of the method was intentionally not change to allow the diff to be easier.

File: src/games/strategy/engine/framework/GameRunner.java (hunk 9 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Yes, look to HeadlessGameServer.java

File: src/games/strategy/engine/framework/map/download/DownloadRunnable.java (hunk 19 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/map/download/DownloadRunnable.java]: Addressed in: 52392020f428e6b7e93e5d71913c8bfc4038aa6d
Really we should not even have the 'downloads' variable to begin with.

File: src/games/strategy/engine/framework/GameRunner.java (hunk 22 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Addressed with 5e47d1a (see commit comment there pls)
Why change it to a foreach when we can just delete the method :boom:

File: src/games/strategy/engine/framework/GameRunner.java (hunk 27 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: I believe you mean it should *not* be an unchecked String. I very much agree that is quite bad. Addressed in: 37e1d96c78dbfe01b595ff199bce3b0762205f64

File: src/games/strategy/engine/framework/GameRunner.java (hunk 24 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Yes, but not going to change right now as part of this PR. The code was imported wholesale from one class to another to allow for easier diffing. Some obvious next steps are to fix up this code, there is much to fix. It's not within the scope of this PR. 

Given the move of code, a good question would be if GameRunner is a better home for this code at this point in time? IMO yes as demonstrated by the variable dependency graph, merging the two allowed variables to go from public or protected to private.

File: src/games/strategy/engine/framework/GameRunner.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: 2 other benefits:
- type safe, do not suffer from incorrect param ordering problem, ie: "doSomething(boolean isHeadless, boolean isSomeOtherOption)"
- if necessary, allows for a third option rather than just a binary

File: src/games/strategy/engine/framework/system/HttpProxy.java (hunk 32 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/system/HttpProxy.java]: Only moving code here. I'm very willing to consider PRs into this branch to further improve things

**Comment by DanVanAtta** [comment]
Removing from milestone, I hacked in a bot map folder command line arg in #1060, so this is cleanup and some potential fixes.

**Comment by DanVanAtta** [comment]
Cutting refactor here. Please review the refactors performed, not the original code.

**Comment by DanVanAtta** [comment]
On second thought, there is some good feedback. gonna keep the ball in my court for a bit longer.

**Comment by DanVanAtta** [comment]
Some pretty mean merge conflicts. The diff is likely not very pretty anymore.. Let's cut the scope on the PR to what it was originally. I will address any commentary relating to the specific refactors done, but will not for any code that was simply moved or otherwise not the target of a refactor.

**Comment by DanVanAtta** [comment]
@RoiEXLab  anything outstanding?
Would like to get this merged in, the most recent round of conflicts were pretty painful.

**Comment by RoiEXLab** [comment]
@DanVanAtta I don't think so, feel free to merge.

**Comment by DanVanAtta** [comment]
@RoiEXLab , turns out this PR was already in flight with a logging fix that would have covedered:  https://github.com/triplea-game/triplea/pull/1141   (https://github.com/triplea-game/triplea/pull/1046/files#diff-c0e299facb8575a7c4ef6eed70c70290R46)
